### PR TITLE
[BIOMAGE-1764] - Fix Error when going back from single-sample DE

### PR DESCRIPTION
**Pull request by aerlaut**
# Description
#### Issue
Clicking "Go Back" from the DE table causes an error to happen. This happens in the piece of code to automatically select the only sample available for the dataset.  The piece of code does dispatch action to update the selected sample, but uses direct assignment, which causes the error.

#### Solution
Dispatch action to update the selected sample correctly.

#### To test
To reproduce the bug :

- Launch a project with a single sample
- Go to Data Management, launch DE with any options
- Click "Go Back"
- This should throw an error, causing an "it's not you, it's us" error page to show up.

# Details
#### URL to issue
https://biomage.atlassian.net/browse/BIOMAGE-1764

#### Link to staging deployment URL (or set N/A)
https://ui-agi-ui655.scp-staging.biomage.net/

#### Links to any PRs or resources related to this PR
N/A

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [X] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [X] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [X] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [x] Unit tests written **or** no unit tests required for change, e.g. documentation update.

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [ ] Started end-to-end tests on the latest commit.

### Documentation updates
- [X] Relevant Github READMEs updated **or** no GitHub README updates required.
- [X] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [ ] Photo of a cute animal attached to this PR.

### CODE REVIEW

**Comment by cosa65** [comment]
>  I'll verify with Vicky about this
This you get the chance to check with @vickymorrison on this?

File: src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx [hunk 1]
```diff
@@ -30,7 +30,7 @@ const DiffExprCompute = (props) => {
   const dispatch = useDispatch();
   const { properties, hierarchy } = useSelector(getCellSets());
   const [isFormValid, setIsFormValid] = useState(false);
-  const [numSamples, setNumSamples] = useState(1)
+  const [numSamples, setNumSamples] = useState(0);
```
Comment (cosa65) [inline src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx:32]: Why was this changed? (and do you know why was it 1 before?)

File: src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx [hunk 2]
```diff
@@ -46,40 +46,30 @@ const DiffExprCompute = (props) => {
...
-
-        dispatch(
-          setComparisonGroup({
-            type,
-            ...comparisonGroup[type],
-            ...deleteKeys,
-          }),
-        );
-      }
+    Object.keys(comparisonGroup).forEach(
```
Comment (cosa65) [inline src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx]: I don't understand too much, but I am guessing that this is dealing with what happens when one of the selected cellSets is deleted, right?

If that is the case, I don't think this should be dealt with here. It should be dealt with the tools we have in redux directly instead of having this go through a different component.

This is effectively dealing with the consequences of a cellSet being deleted, the tool to use for that is a reducer on the `CELL_SETS_DELETE` action. Right now this `useEffect` is performing the work of a reducer in an indirect way.

We trigger the action `CELL_SETS_DELETE`, as a consequence the reducer of each slice should perform its necessary changes.
In this case, I would add a `cellSetsDelete` reducer in the `differentialExpression` slice of the store that checks whether the selected option shoudlb e unselected, and all of this `useEffect` just goes away.

File: src/redux/actions/differentialExpression/loadDifferentialExpression.js [hunk 3]
```diff
@@ -5,14 +5,14 @@ import {
...
-const getCellSetName = (name) => (name?.split('/')[1] || name);
+import { getCellSetKey } from 'utils/cellSets';
 
 const generateDiffExprBody = (experimentId, comparisonGroup, comparisonType, extras) => ({
   name: 'DifferentialExpression',
   experimentId,
-  cellSet: getCellSetName(comparisonGroup.cellSet),
-  compareWith: getCellSetName(comparisonGroup.compareWith),
-  basis: getCellSetName(comparisonGroup.basis),
+  cellSet: getCellSetKey(comparisonGroup.cellSet),
```
Comment (cosa65) [inline src/redux/actions/differentialExpression/loadDifferentialExpression.js:13]: nicee

File: src/utils/cellSets/getCellSetNames.js [hunk 4]
```diff
@@ -0,0 +1,7 @@
+const getCellSetKey = (name) => (name?.split('/')[1] || name);
+const getRootKey = (name) => name?.split('/')[0];
+
+export {
+  getCellSetKey,
+  getRootKey,
+};
```
Comment (cosa65) [inline src/utils/cellSets/getCellSetNames.js]: This file is called `getCellSetNames`, where is this format of cell set name defined? Is it used only in differential expression? Do you know why does this format exist?

Comment (aerlaut) [reply src/utils/cellSets/getCellSetNames.js]: Hmm... yea, maybe the file name should be `getCellSetKeys`. I'll change it.

File: src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx (hunk 1 above)
Comment (aerlaut) [reply src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx:32]: It was set to 1 before because we assume that there will always be at least 1 sample. That is a correct assumption of course, but we have some behaviors which will trigger specifically when the number of samples is 1, even while the cell set is still loading. After the cell set is loaded, the number of samples might change. This means we'll always trigger the behaviour specific for 1 sample, even though in reality there are more than 1 samples. So, I think the better option is to set the initial number of cellsets to 0, which will then be corrected after the cell sets load.

**Comment by aerlaut** [comment]
> > I'll verify with Vicky about this
> > This you get the chance to check with @vickymorrison on this?

Yes I did. We do not want to remove the behaviour, so I implemented the fix.

File: src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx (hunk 2 above)
Comment (aerlaut) [reply src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx]: Great idea! I'll implement that.

File: src/utils/cellSets/getCellSetNames.js (hunk 4 above)
Comment (aerlaut) [reply src/utils/cellSets/getCellSetNames.js]: The format of the cell set name is 'cellClassKey/cellSetKey'. I've adjusted the function name accordingly and put in comment too.

File: src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx (hunk 2 above)
Comment (aerlaut) [reply src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx]: implemented

File: src/redux/reducers/differentialExpression/index.js [hunk 5]
```diff
@@ -39,6 +41,9 @@ const differentialExpressionReducer = (state = initialState, action) => {
     case EXPERIMENT_SETTINGS_PIPELINE_START: {
       return initialState;
     }
+    case CELL_SETS_DELETE: {
+      return differentialExpressionUnselectDeletedOption(state, action);
```
Comment (cosa65) [inline src/redux/reducers/differentialExpression/index.js]: The naming we are using for reducers mean this should just be `cellSetsDelete`

File: src/__test__/components/data-exploration/differential-expression-tool/DiffExprCompute.test.jsx [hunk 6]
```diff
@@ -334,4 +334,66 @@ describe('DiffExprCompute', () => {
       expect(screen.getByText(/Compute/i).closest('button')).not.toBeDisabled();
     });
   });
+
+  it('Unselect deleted option', async () => {
```
Comment (cosa65) [inline src/__test__/components/data-exploration/differential-expression-tool/DiffExprCompute.test.jsx]: Does this test make sense now that we have removed the code handling deletion on the react component?

File: src/redux/reducers/differentialExpression/differentialExpressionUnselectDeletedOption.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+/* eslint-disable no-param-reassign */
+import produce from 'immer';
+import { getCellSetKey } from 'utils/cellSets';
+
+const differentialExpressionUnselectDeletedOption = produce((draft, action) => {
+  const { key: deletedKey } = action.payload;
+
+  Object.keys(draft.comparison.group).forEach((comparisonType) => {
+    const comparisonGroup = draft.comparison.group[comparisonType];
```
Comment (cosa65) [inline src/redux/reducers/differentialExpression/differentialExpressionUnselectDeletedOption.js]: Does this work? usually I can't pick up stuff from `draft` without first doing a `current` or `original`

File: src/redux/reducers/differentialExpression/index.js (hunk 5 above)
Comment (cosa65) [reply src/redux/reducers/differentialExpression/index.js]: Think of it as one single reducer being split up across slices

Comment (aerlaut) [reply src/redux/reducers/differentialExpression/index.js]: Ahh... ok

File: src/__test__/components/data-exploration/differential-expression-tool/DiffExprCompute.test.jsx (hunk 6 above)
Comment (aerlaut) [reply src/__test__/components/data-exploration/differential-expression-tool/DiffExprCompute.test.jsx]: Yes, the test still stand, because it's testing for the expected outcome, not the functionality.

File: src/redux/reducers/differentialExpression/differentialExpressionUnselectDeletedOption.js (hunk 7 above)
Comment (cosa65) [inline src/redux/reducers/differentialExpression/differentialExpressionUnselectDeletedOption.js]: This could be replaced with 
`Object.entries(draft.comparison.group).forEach(([comparisonType, comparisonGroup]) => {`

File: src/redux/reducers/differentialExpression/differentialExpressionUnselectDeletedOption.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+    Object.keys(comparisonGroup).forEach((comparisonKey) => {
+      if (getCellSetKey(comparisonGroup[comparisonKey]) !== deletedKey) return;
```
Comment (cosa65) [inline src/redux/reducers/differentialExpression/differentialExpressionUnselectDeletedOption.js]: Here there's the same pattern of using `Object.keys` but then getting the value instead of using `Object.entries` directly

File: src/redux/reducers/differentialExpression/differentialExpressionUnselectDeletedOption.js (hunk 7 above)
Comment (aerlaut) [reply src/redux/reducers/differentialExpression/differentialExpressionUnselectDeletedOption.js]: Yea, it seems to work... I see the intended outcome. I'll check if there is a difference between using `current()` or direct access like this.

File: src/__test__/components/data-exploration/differential-expression-tool/DiffExprCompute.test.jsx (hunk 6 above)
Comment (cosa65) [reply src/__test__/components/data-exploration/differential-expression-tool/DiffExprCompute.test.jsx]: So it is testing the redux action?

File: src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx (hunk 1 above)
Comment (cosa65) [reply src/components/data-exploration/differential-expression-tool/DiffExprCompute.jsx:32]: nice

File: src/__test__/components/data-exploration/differential-expression-tool/DiffExprCompute.test.jsx (hunk 6 above)
Comment (aerlaut) [reply src/__test__/components/data-exploration/differential-expression-tool/DiffExprCompute.test.jsx]: Yes

File: src/redux/reducers/differentialExpression/differentialExpressionUnselectDeletedOption.js (hunk 7 above)
Comment (aerlaut) [reply src/redux/reducers/differentialExpression/differentialExpressionUnselectDeletedOption.js]: So it seems that `current()` has to be used on a `draft`, which is a proxy to an object. When you directly access a value that is not an object, accessing draft directly returns the value, so it's comparable. If you try to access `current()` on a primtive value (e.g. not-object), it'll throw an error.

In this case, I'm accessing a property which is a primitive(string), so it returns the real value which can then be used to compare.

Comment (aerlaut) [reply src/redux/reducers/differentialExpression/differentialExpressionUnselectDeletedOption.js]: Ultimately used `Object.values` because we don't need `comparisonType`.

File: src/redux/reducers/differentialExpression/index.js (hunk 5 above)
Comment (aerlaut) [reply src/redux/reducers/differentialExpression/index.js]: reducer renamed

File: src/redux/reducers/differentialExpression/cellSetsDelete.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
+/* eslint-disable no-param-reassign */
+import produce from 'immer';
+import { getCellSetKey } from 'utils/cellSets';
+
+const differentialExpressionUnselectDeletedOption = produce((draft, action) => {
+  const { key: deletedKey } = action.payload;
+
+  Object.values(draft.comparison.group).forEach((comparisonGroup) => {
+    Object.keys(comparisonGroup).forEach((comparisonKey) => {
+      if (getCellSetKey(comparisonGroup[comparisonKey]) !== deletedKey) return;
```
Comment (cosa65) [inline src/redux/reducers/differentialExpression/cellSetsDelete.js]: This one wasn't changed for this comment yet, right? https://github.com/hms-dbmi-cellenics/ui/pull/655#discussion_r829051914

Comment (aerlaut) [reply src/redux/reducers/differentialExpression/cellSetsDelete.js]: I changed it in a378b88 to use

    Object.values(draft.comparison.group).forEach((comparisonGroup) => { ...

**Review by cosa65 — APPROVED** [review]
